src/main/java/org/repodriller/RepoVisitor.java

```diff
@@ -0,0 +1,168 @@
+	}
+
+	/**
+	 * Print a description of all of the CommitVisitors I maintain.
+	 */
+	void printVisitors() {
+		for(CommitVisitor visitor : visitors.keySet()) {
+			log.info("- " + visitor.name() + " (" + visitor.getClass().getName() + ")");
+		}
+	}
+
+	/**
+	 * @return True if no CommitVisitors, else false.
+	 */
+	public boolean isEmpty() {
+		return visitors.isEmpty();
+	}
+
+	/***
+	 * Methods for visiting an SCMRepository.
+	 ***/
+
+	/**
+	 * Prepare for the next repo we visit.
+	 * Calls {@link CommitVisitor#initialize} on each CommitVisitor in my collection.
+	 *
+	 * @param repo	The repo we are visiting next
+	 */
+	void beginRepoVisit(SCMRepository repo) {
+		if (currentRepo != null)
+			throw new RepoDrillerException("Error, one repo at a time. We are already visiting a repo.");
+		currentRepo = repo;
+
+		for(Map.Entry<CommitVisitor, PersistenceMechanism> entry : visitors.entrySet()) {
+			CommitVisitor visitor = entry.getKey();
+			PersistenceMechanism writer = entry.getValue();
+
+			try {
+				log.info("-> Initializing visitor " + visitor.name());
+				visitor.initialize(currentRepo, writer);
+			} catch (Exception e) {
+				log.error("error in " + currentRepo.getPath() +
+						"when initializing " + visitor.name() + ", error=" + e.getMessage(), e);
+			}
+		}
+	}
+
+	/**
+	 * Visit this commit in the current repo.
+	 * Calls {@link CommitVisitor#process} on each CommitVisitor in my collection.
+	 *
+	 * @param commit
+	 */
+	void visitCommit(Commit commit) {
+		for(Map.Entry<CommitVisitor, PersistenceMechanism> entry : visitors.entrySet()) {
+			CommitVisitor visitor = entry.getKey();
+			PersistenceMechanism writer = entry.getValue();
+
+			try {
+				log.info("-> Processing " + commit.getHash() + " with " + visitor.name());
+				visitor.process(currentRepo, commit, writer);
+			} catch (CSVFileFormatException e) {
+				log.fatal(e);
+				System.exit(-1);
+			} catch (Exception e) {
+				log.error("error processing #" + commit.getHash() + " in " + currentRepo.getPath() +
+						", processor=" + visitor.name() + ", error=" + e.getMessage(), e);
+			}
+		}
+	}
+
+	/**
+ 	 * Conclude the visit to the current repo.
+ 	 * Calls {@link CommitVisitor#finalize} on each CommitVisitor in my collection.
+	 */
+	void endRepoVisit() {
+		for(Map.Entry<CommitVisitor, PersistenceMechanism> entry : visitors.entrySet()) {
+			CommitVisitor visitor = entry.getKey();
+			PersistenceMechanism writer = entry.getValue();
+
+			try {
+				log.info("-> Finalizing visitor " + visitor.name());
+				visitor.finalize(currentRepo, writer);
+			} catch (Exception e) {
+				log.error("error in " + currentRepo.getPath() +
+						"when finalizing " + visitor.name() + ", error=" + e.getMessage(), e);
+			}
+		}
+
+		currentRepo = null;
+	}
+
+	/**
+	 * Close all of the PersistenceMechanism's associated with visitors.
+	 */
+	void closeAllPersistence() {
+		/* TODO: This API doesn't make sense to me. Why are the persistence mechanisms closed by this class?
```

Comment:
Where would you close them? Would you leave for the client to close them?

The problem is that, then, the user can't write a beautiful fluent instantiation of the repo:

```
new RepositoryMining().
...
.process(visitor, new CSVFile(...))
```

The user would have to store `new CSVFile()` in a variable, and then, close it. 

In addition, I'd not expect the program to do anything else after repodriller is done. So, I guess I implemented it as a 'courtesy' to the user.

What do you think?

Reply:
I agree that the inline PersistenceMechanism instantiation is lovely, and that if code is written in the functional style then RepositoryMining must close up the PersistenceMechanism's it's been given.

On the other hand, there's a mismatch between where the PM is created (client level) and where it is closed (framework level), which is surprising.

How about we overload `RepositoryMining.process` with a "close the PM for me" flag, and have the current 2-arg version always close the PM?

Reply:
Let's open an issue to discuss about that (#77), so that I can move on with this PR